### DISCUSSION TITLE
`FullTensorProduct` in relation to CGC table

### DISCUSSION THREAD

**Discussion by JSLJ23** [e3nn/e3nn#422] [Q&A]
```python
import torch
from e3nn import o3

x = torch.tensor ([1.0, 2.0, 3.0])
y = torch.tensor ([4.0, 5.0, 6.0])

vector_irreps = o3.Irreps("1x1o")

tp = o3.FullTensorProduct(vector_irreps, vector_irreps)
print(tp)
> FullTensorProduct(1x1o x 1x1o -> 1x0e+1x1e+1x2e | 3 paths | 0 weights)

tp(x, y)
> tensor([18.4752, -2.1213,  4.2426, -2.1213, 12.7279,  9.1924, -0.8165, 19.0919, 9.8995])
```

Hi e3nn developers,
I am trying to understand what is going on under the hood of `FullTensorProduct` when it performs the tensor product on x and y. I can see that the first element of the output is essentially the dot product of x and y scaled by 1/sqrt(3), while the next 3 elements are the cross product scaled by 1/sqrt(2). I was wondering how does the `FullTensorProduct` get the right combinations of elements to multiply to product the l-0, 1 and 2 irrep in the resultant tensor?

Example of computing the J=0, M=0 from the CGC table given `x` and `y`:
```python
from math import sqrt
1.0 * 6.0 * sqrt(1/3) + 2.0 * 5.0 * -sqrt(1/3) + 3.0 * 4.0 * sqrt(1/3)
> 4.618802153517006
```

I was referring to a traditional CGC table and it seems that when I manually try to compute the terms from that table following the ordering of m1 and m2 to get M in the result, the values don't match what is computed via `tp(x, y)`. Was hoping to get some guidance on this to understand how the `FullTensorProduct` relates to the tensor product computation on CGC tables.

**Comment by mariogeiger** [comment]
This is the most common question we get here :)

Short answer: Our math is correct, This is just a matter of convention (choice of basis for our irreps).

See for instance
https://github.com/e3nn/e3nn/discussions/419
https://github.com/e3nn/e3nn/discussions/405
https://docs.e3nn.org/en/latest/api/o3/o3_wigner.html#e3nn.o3.wigner_3j
https://docs.e3nn.org/en/latest/guide/change_of_basis.html

**Reply by JSLJ23** [reply]
Sorry I don't fully understand (or maybe I am missing something) the reply in https://github.com/e3nn/e3nn/discussions/405.

1. Is there a general rule for this change of basis?
2. Is there a reason for it too?
3. And for the sign difference in the 2nd row, is this also accounted for?

**Reply by mariogeiger** [reply]
1. https://github.com/e3nn/e3nn-jax/blob/main/e3nn_jax/_src/so3.py#L6 https://en.wikipedia.org/wiki/Spherical_harmonics#Real_form
2. We prefer real numbers over one generator being diagonal 
3. Can you elaborate on that?

more on 1:
if you install `e3nn-jax` you can print the complex and the real CGs:
```
import e3nn_jax as e3nn
print(e3nn.clebsch_gordan(1, 1, 1))
print(e3nn.su2_clebsch_gordan(1, 1, 1))
```
